damus/Models/NoteContent.swift

```diff
@@ -73,83 +73,89 @@ func render_note_content(ev: NostrEvent, profiles: Profiles, keypair: Keypair) -
+            // If there are previewable blocks that occur before the consecutive sequence of them at the end of the content,
+            // we should not hide the text representation of any previewable block to avoid altering the format of the note.
+            if block.is_previewable && ind < hide_text_index {
+                hide_text_index = blocks.endIndex
+            }
+
+            // No need to show the text representation of the block if the only previewables are the sequence of them
+            // found at the end of the content.
+            // This is to save unnecessary use of screen space.
+            if ind >= hide_text_index {
                 return str
             }
+        }
+
+        switch block {
+        case .mention(let m):
             return str + mention_str(m, profiles: profiles)
         case .text(let txt):
-            return str + CompatibleText(stringLiteral: reduce_text_block(blocks: blocks, ind: ind, txt: txt, one_note_ref: one_note_ref))
-
+            // Trim trailing whitespace if the following blocks will be hidden.
+            if ind == hide_text_index - 1 {
+                return str + CompatibleText(stringLiteral: trim_suffix(txt))
+            } else {
+                return str + CompatibleText(stringLiteral: txt)
+            }
         case .relay(let relay):
             return str + CompatibleText(stringLiteral: relay)
-            
         case .hashtag(let htag):
             return str + hashtag_str(htag)
         case .invoice(let invoice):
-            invoices.append(invoice)
-            return str
+            return str + invoice_str(invoice)
         case .url(let url):
-            let url_type = classify_url(url)
-            switch url_type {
-            case .media:
-                urls.append(url_type)
-                return str
-            case .link(let url):
-                urls.append(url_type)
-                return str + url_str(url)
-            }
+            return str + url_str(url)
         }
     }
 
     return NoteArtifactsSeparated(content: txt, words: bs.words, urls: urls, invoices: invoices)
 }
 
-func reduce_text_block(blocks: [Block], ind: Int, txt: String, one_note_ref: Bool) -> String {
-    var trimmed = txt
-    
-    if let prev = blocks[safe: ind-1],
-       case .url(let u) = prev,
-       classify_url(u).is_media != nil {
-        trimmed = " " + trim_prefix(trimmed)
-    }
-    
-    if let next = blocks[safe: ind+1] {
-        if case .url(let u) = next, classify_url(u).is_media != nil {
-            trimmed = trim_suffix(trimmed)
-        } else if case .mention(let m) = next,
-                  case .note = m.ref,
-                  one_note_ref {
-            trimmed = trim_suffix(trimmed)
-        }
-    }
-    
-    return trimmed
+func invoice_str(_ invoice: Invoice) -> CompatibleText {
+    // Add foreground color to Lightning invoice but no need to link to anything as the attached preview is sufficient.
```

Comment:
Oops, this comment is wrong. I forgot to remove it after I changed my approach mid-way through this work.

Reply:
Removed in 82fcf123d485cf016ab64a1bea8299b022d44711